Add Get-CVClientLicense/Revoke-CVClientLicense function

CraigTolley opened the pull request:
Hi,

Have added a new function to specifically get the client licenses. Am working on a counter function to remove client licenses as well.

Hope format is ok. Tried to keep with everything that you have already done. 

Craig

Commit:
Add Get-ClientLicense function

Commit:
Remove incorrect additional parameters

Commit:
Add Revoke-CVClientLicense function

CraigTolley commented:
#4

gstoops commented on Modules/Commvault.CommCell/Commvault.CommCell.psm1:
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
+    Author: Craig Tolley
+    Based on API Documentation at https://documentation.commvault.com/commvault/v11/article?p=47685.htm
+#>
+    [CmdletBinding(DefaultParameterSetName = 'Default')]
```
default should be 'ByName'

gstoops commented on Modules/Commvault.CommCell/Commvault.CommCell.psm1:
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
+    [OutputType([PSCustomObject])]
+    param(
+        [Alias('Client')]
+        [Parameter(Mandatory = $False, ParameterSetName = 'ByName')]
```
[Parameter(Mandatory = $True, ParameterSetName = 'ByName', ValueFromPipeline = $True, ValueFromPipelineByPropertyName = $True)]

gstoops commented on Modules/Commvault.CommCell/Commvault.CommCell.psm1:
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
+        [String] $Name,
+
+        [Alias('ClientId')]
+        [Parameter(Mandatory = $False, ParameterSetName = 'ById')]
```
[Parameter(Mandatory = $True, ParameterSetName = 'ByName', ValueFromPipeline = $True, ValueFromPipelineByPropertyName = $True)]

gstoops commented on Modules/Commvault.CommCell/Commvault.CommCell.psm1:
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
+            }
+            else {
+                $foundClient = $null
+            }
```
Delete these lines.

gstoops commented on Modules/Commvault.CommCell/Commvault.CommCell.psm1:
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
+
+            if ($PSCmdlet.ParameterSetName -eq 'ByName') {
+                $Id = (Get-CVClient -Name $Name).clientId
+            }
```
if ($PSCmdlet.ParameterSetName -eq 'ByName') {
                $clientObj = (Get-CVClient -Name $Name)
                if ($null -ne $clientObj) {
                    $Id = $clientObj.clientId
                }
                else {
                    Write-Information -InformationAction Continue -MessageData "INFO: $($MyInvocation.MyCommand): client not found having name [$Name]"
                    return
                }
            }

gstoops commented on Modules/Commvault.CommCell/Commvault.CommCell.psm1:
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
+            }
+
+            $sessionObj.requestProps.endpoint = $sessionObj.requestProps.endpoint -creplace ('{clientId}', $Id)
+            $headerObj = Get-CVRESTHeader $sessionObj
```
please add blank line here.

gstoops commented on Modules/Commvault.CommCell/Commvault.CommCell.psm1:
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
+    
+            $response = Submit-CVRESTRequest $payload $validate
+    
+            if ($response.IsValid) {
+                Write-Output $response.Content.licensesInfo
+            }
```
if ($response.IsValid) {
                Write-Output $response.Content.licensesInfo
            }
            else {
                Write-Information -InformationAction Continue -MessageData "INFO: $($MyInvocation.MyCommand): license information not found for client Id [$Id]"
            }

gstoops commented on Modules/Commvault.CommCell/Commvault.CommCell.psm1:
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
+
+        [Parameter(Mandatory = $True, ParameterSetName = 'ByName')]
+        [Parameter(Mandatory = $True, ParameterSetName = 'ById')]
+        $LicensesToRelease
```
What type is this?

gstoops commented on Modules/Commvault.CommCell/Commvault.CommCell.psm1:
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
+    Based on API Documentation at https://documentation.commvault.com/commvault/v11/article?p=92013.htm
+#>
+    Param (
+        [Alias('ClientName')]
```
The $Name and $Id params here should be coded identical to the Get function

gstoops commented on Modules/Commvault.CommCell/Commvault.CommCell.psm1:
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
+    Author: Craig Tolley
+    Based on API Documentation at https://documentation.commvault.com/commvault/v11/article?p=92013.htm
+#>
+    Param (
```
[CmdletBinding(DefaultParameterSetName = 'ByName', SupportsShouldProcess = $True, ConfirmImpact = 'High')]

gstoops commented on Modules/Commvault.RESTSession/Commvault.RESTSession.psm1:
```diff
@@ -600,6 +599,22 @@ function GetAPIDetail ([String] $Request) {
+                Body        = ''
+            
+            }
             'Add-CVClientGroup' = @{
```
Please add blank line here.

gstoops reviewed (changes requested):
Hi Craig, great work here. We just need to address a few things. :)
Please push your changes to Commvault.CommCell.psd1 for the new function exports:
FunctionsToExport = @('Get-CVAlert', 'Get-CVClientGroup', 'Get-CVClient', 'Get-CVClientLicense', 'Revoke-CVClientLicense',
                      'Get-CVSubclient', 'Get-CVVersionInfo', 'Set-CVClient', 'Set-CVClientGroup')

Commit:
Fixes

CraigTolley commented:
All suggestions incorporated.

CraigTolley commented:
Is there anything else that you need me to do before this can get merged in?

Commit:
Add Enable/Disable-CVClientActivity functions

New functions to allow disabling of activities on Clients.
New enum CVActivityType to limit options

CraigTolley commented:
Closed as updating my repos and creating a new PR
